nixos/modules/services/web-apps/immich.nix

```
@@ -0,0 +1,247 @@
+{ config, lib, pkgs, self, ... }:
+let
+  cfg = config.services.immich;
+  isUnixSocket = lib.hasPrefix "/" cfg.database.host;
+  postgresEnv = if isUnixSocket then {
+    DB_URL = "socket://${cfg.database.host}?dbname=${cfg.database.name}";
+  } else {
+    DB_HOSTNAME = cfg.database.host;
+    DB_PORT = toString cfg.database.port;
+    DB_DATABASE_NAME = cfg.database.name;
+    DB_USERNAME = cfg.database.user;
+  };
+
+  commonServiceConfig = {
+    Type = "simple";
+    Restart = "on-failure";
+    RestartSec = 3;
+
+    # Hardening
+    CapabilityBoundingSet = "";
+    NoNewPrivileges = true;
+    PrivateUsers = true;
+    PrivateTmp = true;
+    PrivateDevices = true;
+    PrivateMounts = true;
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    RestrictAddressFamilies = [
+      "AF_INET"
+      "AF_INET6"
+      "AF_UNIX"
+    ];
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    RestrictSUIDSGID = true;
+  };
+  inherit (lib) types mkIf mkOption mkEnableOption;
+in {
+  options.services.immich = {
+    enable = mkEnableOption "Immich";
+    package = lib.mkPackageOption pkgs "immich" {};
+
+    mediaLocation = mkOption {
+      type = types.path;
+      default = "/var/lib/immich";
+      description = "Directory used to store media files. If this directory already exists the immich user has to be able to read and write to it.";
```

```suggestion
      description = "Directory used to store media files. If it is not the default the immich user has to be able to read and write to the directory.";
```
Only `/var/lib/immich` is created via `serviceConfig.StateDirectory`.

---

I'd even say using `systemd.tmpfiles.settings` in the module would be more fool-proof.

---

What's wrong with `StateDirectory`?

---

The StateDirectory is always relative to `/var/lib/`.

---

And that's where state belongs.

---

In that case you'd need an option like this one to make the service use it though, no?

I agree that the service shouldn't create it though, that should be explicitly left up to the user and communicated as such.